hack/generated/controllers/generic_controller.go

```diff
@@ -0,0 +1,749 @@
+			err := gr.ARMClient.DeleteDeployment(ctx, deployment.Id)
+			if err != nil {
+				return errors.Wrapf(err, "failed to delete deployment %q", deployment.Id)
+			}
+			deployment.Id = ""
+			deployment.Name = ""
+		}
+
+		err = gr.updateMetaObject(data.metaObj, deployment, status)
+		if err != nil {
+			return errors.Wrap(err, "failed updating metaObj")
+		}
+
+		return nil
+	}); err != nil {
+		// This is a superfluous error as per https://github.com/kubernetes-sigs/controller-runtime/issues/377
+		// The correct handling is just to ignore it and we will get an event shortly with the updated version to patch
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+
+		return ctrl.Result{}, errors.Wrap(err, "failed to patch")
+	}
+
+	result := ctrl.Result{}
+	// TODO: This is going to be common... need a wrapper/helper somehow?
+	if !deployment.IsTerminalProvisioningState() {
+		result = ctrl.Result{
+			RequeueAfter: 5 * time.Second,
+		}
+	}
+	return result, err
+}
+
+func (gr *GenericReconciler) resourceSpecToDeployment(data *ReconcileMetadata) (*armclient.Deployment, error) {
+	resourceGroupName, typedArmSpec, err := ResourceSpecToArmResourceSpec(gr, data.metaObj)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: get other deployment details from status and avoid creating a new deployment
+	deploymentId, deploymentIdOk := data.metaObj.GetAnnotations()[DeploymentIdAnnotation]
+	deploymentName, deploymentNameOk := data.metaObj.GetAnnotations()[DeploymentNameAnnotation]
+	if deploymentIdOk != deploymentNameOk {
+		return nil, errors.Errorf(
+			"deploymentIdOk: %t, deploymentNameOk: %t expected to match, but didn't",
+			deploymentIdOk,
+			deploymentNameOk)
+	}
+
+	var deployment *armclient.Deployment
+	if deploymentIdOk && deploymentNameOk {
+		deployment = gr.ARMClient.NewDeployment(resourceGroupName, deploymentName, typedArmSpec)
+		deployment.Id = deploymentId
+	} else {
+		deploymentName, err := CreateDeploymentName()
+		if err != nil {
+			return nil, err
+		}
+		deployment = gr.ARMClient.NewDeployment(resourceGroupName, deploymentName, typedArmSpec)
+	}
+	return deployment, nil
+}
+
+func hasResourceHashAnnotationChanged(metaObj genruntime.MetaObject) (bool, error) {
+	oldSig, exists := metaObj.GetAnnotations()[ResourceSigAnnotationKey]
+	if !exists {
+		// signature does not exist, so yes, it has changed
+		return true, nil
+	}
+
+	newSig, err := SpecSignature(metaObj)
+	if err != nil {
+		return false, err
+	}
+	// check if the last signature matches the new signature
+	return oldSig != newSig, nil
+}
+
+func getResourceProvisioningState(metaObj genruntime.MetaObject) *armclient.ProvisioningState {
+	state, ok := metaObj.GetAnnotations()[ResourceStateAnnotation]
+	if !ok {
+		return nil
+	}
+
+	result := armclient.ProvisioningState(state)
+	return &result
+}
+
+func getDeploymentId(metaObj genruntime.MetaObject) string {
+	id, ok := metaObj.GetAnnotations()[DeploymentIdAnnotation]
```

Comment:
Would it be better to add a `GetAnnotation(string)` method that just returns the one you want?

Reply:
It's impl would just be: 
```
func getAnnotation(metaObj genruntime.MetaObject, annotation string) (string, bool) {
    return metaObj.GetAnnotations()[annotation]
}
```

which doesn't seem to add much value. If we wanted to obscure the source of the deployment id (is it in the annotations dictionary, on status, etc?) then I think having this "getter" makes more sense. The getter knowing that yes indeed `deploymentId` is in the annotations dictionary makes sense I think?

Plus a getter per-prop means that we can have non-string returns, which in some cases we need (see `getShouldPreserveDeployment`).